src/devtools/client/debugger/src/components/Editor/Breakpoints/Panel/PanelEditor.js

```diff
@@ -0,0 +1,190 @@
+import React, { useEffect, useState, PureComponent } from "react";
+import classnames from "classnames";
+import { connect } from "devtools/client/debugger/src/utils/connect";
+import { getContext } from "devtools/client/debugger/src/selectors";
+import { PanelInput } from "./PanelInput";
+import actions from "devtools/client/debugger/src/actions";
+
+class PanelEditor extends PureComponent {
```

Comment:
Used the editor name here instead since it's more appropriate.

Reply:
This is intentionally kept as a class component, because using a functional component with the current implementation of PanelInput doesn't work out very well. See [stale props inside functional components](https://reactjs.org/docs/hooks-faq.html#why-am-i-seeing-stale-props-or-state-inside-my-function).

We only assign the event listeners (e.g. `handleKeyDown`) to PanelInput when it mounts. The problem with functional components is that a callback like `handleKeyDown` “sees” the props and state from the render it was created in. In this case, the `handleKeyDown` callback attached to codeMirror when it mounts is never updated, and will keep using the old state/props as if they were never updated. 

For example, editing a logValue from "Hello" to "Hello World" and pressing the Enter key to save the new log message will call `onKeyDown` and `handleSave`. But both will end up referring to stale state, where the logValue is still "Hello". The logValue then ends up being saved as "Hello".

This problem simply goes away if we just use a class component instead.